crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,742 @@
+    current_blank_characters: usize,
+}
+
+impl<'a> LinePreprocessor<'a> {
+    fn new(tokens: &'a [LexResult], locator: &'a Locator) -> LinePreprocessor<'a> {
+        LinePreprocessor {
+            tokens: tokens.iter().flatten(),
+            locator,
+            previous_blank_lines: 0,
+            current_blank_lines: 0,
+            current_blank_characters: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for LinePreprocessor<'a> {
+    type Item = LogicalLineInfo;
+
+    fn next(&mut self) -> Option<LogicalLineInfo> {
+        let mut line_is_comment_only = true;
+        let mut is_docstring = true;
+        let mut first_token: Option<Tok> = None;
+        let mut first_token_range: Option<TextRange> = None;
+        let mut last_token: Option<Tok> = None;
+        let mut parens = 0u32;
+
+        while let Some((token, range)) = self.tokens.next() {
+            if matches!(token, Tok::Indent | Tok::Dedent) {
+                continue;
+            }
+
+            if !matches!(token, Tok::Newline) {
+                // Ideally, we would like to have a "async def" token since we care about the "def" part.
+                // As a work around, we ignore the first token if it is "async".
+                if first_token.is_none() && !matches!(token, Tok::Async) {
+                    first_token = Some(token.clone());
+                }
+                if first_token_range.is_none() {
+                    first_token_range = Some(*range);
+                }
+
+                if !matches!(token, Tok::NonLogicalNewline) {
+                    if !matches!(token, Tok::Comment(_)) {
+                        line_is_comment_only = false;
+                    }
+
+                    // Allow a comment to follow a docstring.
+                    if !matches!(token, Tok::String { .. } | Tok::Comment(_)) {
+                        is_docstring = false;
+                    }
+                }
+
+                last_token = Some(token.clone());
+            }
+
+            match token {
+                Tok::Lbrace | Tok::Lpar | Tok::Lsqb => {
+                    parens = parens.saturating_add(1);
+                }
+                Tok::Rbrace | Tok::Rpar | Tok::Rsqb => {
+                    parens = parens.saturating_sub(1);
+                }
+                Tok::Newline | Tok::NonLogicalNewline if parens == 0 => {
+                    let last_token_range = *range;
+
+                    if !matches!(first_token, Some(Tok::String { .. })) {
+                        is_docstring = false;
+                    }
+
+                    let first_range = first_token_range.unwrap();
+
+                    let range = if matches!(first_token, Some(Tok::Indent)) {
+                        first_range
+                    } else {
+                        TextRange::new(
+                            self.locator.line_start(first_range.start()),
+                            first_range.start(),
+                        )
+                    };
```

Comment:
I just noticed that `expand_indent` hardcodes the size of tabs to 8, which is incorrect. It should respect the `LinterSettings::tab_size` setting

Reply:
This isn't something I added in this PR, so I will open another PR for this if that's ok.

Reply:
PR opened: https://github.com/astral-sh/ruff/pull/9506